### ISSUE TITLE
Force uv to skip certain (transitive) dependencies

### ISSUE THREAD

**Issue by chrisfougner** [astral-sh/uv#9174]
Is there a way to force uv to skip certain dependencies? Specifically we use docker containers [maintained by Nvidia](https://catalog.ngc.nvidia.com/orgs/nvidia/containers/pytorch), which come pre-installed with various python ML packages on the system level (torch, apex, amp_C, flash-attn). If I use `uv venv --system-site-packages` then I can access those dependencies, the problem is that if any package that I install has a transitive dependency on one of these packages (primarily torch), then the one installed in venv will be chosen over the system one, and things break. I'd like to be able to specify that torch for example should not be installed as a transitive dependency, but I'm not sure how to do this.

**Comment by danielhollas** [comment]
It would be great to have a proper solution for this, not just a hack. We have a similar situation in our container setup, but the number of packages much larger.

See also discussion in #2500

**Comment by charliermarsh** [comment]
I think using overrides with an "impossible" marker is actually kind of a fine solution? The alternative is that we introduce something non-standard.

**Comment by charliermarsh** [comment]
I think this should be tracked in #4422 though.

**Comment by chrisfougner** [comment]
Thank you both. Super helpful. The solution seems fine to me.

For anyone else that stubmless upon this issue. Add the following to your pyproject.toml. If you have a top-level pyproject.toml and any additional package specific ones, you should add it to the top-level pyproject.toml

```
[tool.uv]
override-dependencies = [
    "torch; sys_platform == 'never'",
]
```

**Comment by ahakanbaba** [comment]
> Thank you both. Super helpful. The solution seems fine to me.
> 
> For anyone else that stubmless upon this issue. Add the following to your pyproject.toml. If you have a top-level pyproject.toml and any additional package specific ones, you should add it to the top-level pyproject.toml
> 
> ```
> [tool.uv]
> override-dependencies = [
>     "torch; sys_platform == 'never'",
> ]
> ```


This is a very reasonable workaround. 
BUT, this does not work with `--universal` parameter :( 


Instead I have written an obviously false another constraint. For example 

```
[tool.uv]
override-dependencies = [
    "torch; python_version < '3.0'",
]
```

**Comment by atiladeokegab** [comment]
thank you

**Comment by brainbytes42** [comment]
thank you @ahakanbaba for your solution - this saved me while struggeling with keras-hub.

But, what I don't understand, [as I wrote here also](https://github.com/keras-team/keras-hub/issues/2101#issuecomment-3951630054), is why the python_version constraint does work, but neither sys_platform, nor platform_system lets me restrict an import not to be used on windows (neither restrcting it on linus oder negationg for windows).

the dependency [specified over at keras-hub](https://github.com/keras-team/keras-hub/blob/9569150618c5db4252587ab45cf2724fd5b0b874/pyproject.toml#L37) tries to exclude the dependency on windows (there is no tensorflow-text for windows), but this doesn't work for me. 
also, if I add the dependency by hand and specify different versions of markers, e.g. one of
```
platform_system == 'Linux'
platform_system != 'Windows'
sys_platform == 'linux'
```
uv seems to still require the dependency (the keras-hub dependency was removed to test this ;-) ).

For my understanding, it 'should' just work to depend on keras-hub on windows, which would then omit the dependency on tensorflow-text (see link above)...?

(I'm using the current version of uv (0.10.5) on windows 11.)

Thanks for any hints on this topic, as I'm quite confused whether this is a uv- and/or keras-hub issue or just 'best of both worlds clashing'... ;-)
